sdl_viewer/src/graphic/program.rs

```diff
@@ -0,0 +1,141 @@
+        }
+        GlProgram { gl: self.gl, id }
+    }
+}
+
+pub struct GlProgram {
+    pub gl: Rc<opengl::Gl>,
+    pub id: GLuint,
+}
+
+impl Drop for GlProgram {
+    fn drop(&mut self) {
+        unsafe {
+            self.gl.DeleteProgram(self.id);
+        }
+    }
+}
+
+fn compile_shader(gl: &opengl::Gl, code: &str, kind: GLenum) -> GLuint {
+    let shader;
+    unsafe {
+        shader = gl.CreateShader(kind);
+        let c_str = CString::new(code.as_bytes()).unwrap();
+        gl.ShaderSource(shader, 1, &c_str.as_ptr(), ptr::null());
+        gl.CompileShader(shader);
+        let mut status = i32::from(opengl::FALSE);
+        gl.GetShaderiv(shader, opengl::COMPILE_STATUS, &mut status);
+        if status != (i32::from(opengl::TRUE)) {
+            let mut len = 0;
+            gl.GetShaderiv(shader, opengl::INFO_LOG_LENGTH, &mut len);
+            let mut buf = Vec::with_capacity(len as usize);
+            buf.set_len((len as usize) - 1); // subtract 1 to skip the trailing null character
+            gl.GetShaderInfoLog(
+                shader,
+                len,
+                ptr::null_mut(),
+                buf.as_mut_ptr() as *mut GLchar,
+            );
+            panic!(
+                "{}",
+                str::from_utf8(&buf).expect("ShaderInfoLog invalid UTF8")
+            );
+        }
+    }
+    shader
+}
+
+fn link_program(
+    gl: &opengl::Gl,
+    vertex_shader_id: GLuint,
+    geometry_shader_id: Option<GLuint>,
+    fragment_shader_id: Option<GLuint>,
+) -> GLuint {
+    unsafe {
+        let program = gl.CreateProgram();
+        gl.AttachShader(program, vertex_shader_id);
+        if let Some(id) = geometry_shader_id {
+            gl.AttachShader(program, id)
+        };
+        if let Some(id) = fragment_shader_id {
+            gl.AttachShader(program, id);
+        }
+        gl.LinkProgram(program);
+        gl.DetachShader(program, vertex_shader_id);
+        if let Some(id) = geometry_shader_id {
+            gl.DetachShader(program, id)
+        };
+        if let Some(id) = fragment_shader_id {
+            gl.DetachShader(program, id)
+        };
+
+        let mut status = i32::from(opengl::FALSE);
+        gl.GetProgramiv(program, opengl::LINK_STATUS, &mut status);
+        if status != (i32::from(opengl::TRUE)) {
+            let mut len: GLint = 0;
+            gl.GetProgramiv(program, opengl::INFO_LOG_LENGTH, &mut len);
+            let mut buf = Vec::with_capacity(len as usize);
+            buf.set_len((len as usize) - 1); // subtract 1 to skip the trailing null character
+            gl.GetProgramInfoLog(
+                program,
+                len,
+                ptr::null_mut(),
+                buf.as_mut_ptr() as *mut GLchar,
+            );
+            panic!(
+                "{}",
+                str::from_utf8(&buf).expect("ProgramInfoLog invalid UTF8")
+            );
+        }
+        program
```

Comment:
Maybe not really generic, just a regular function that takes either opengl::LINK_STATUS or opengl::COMPILE_STATUS? Also, ideally, I think it shouldn't panic but return a Result, but I didn't want to make this PR even larger.

Reply:
Actually, I think it might not buy us much to have this function. Fine to leave it like it is? At least it's not making it worse, this is just moved from glhelper.rs :)

Reply:
Yeah, I agree. Even if there is again I don't think it should be in the scope of this PR :)